dll/win32/browseui/shellbars/CBandSiteMenu.cpp

```diff
@@ -103,12 +125,215 @@ HRESULT CBandSiteMenu::CreateMenuPart()
+
+    if (LoadStringW(GetModuleHandleW(L"browseui.dll"), IDS_BROWSEFORNEWTOOLAR, message, _countof(message)))
+        bi.lpszTitle = message;
+    else
+        bi.lpszTitle = L"Choose a folder";
+
+    LPITEMIDLIST pidlSelected = SHBrowseForFolderW(&bi);
+    if (pidlSelected == NULL)
+        return S_FALSE;
+
+    CComPtr<IShellFolderBand> pISFB;
+    HRESULT hr = CISFBand_CreateInstance(IID_IShellFolderBand, (PVOID*)&pISFB);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    hr = pISFB->InitializeSFB(NULL, pidlSelected);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    return pISFB->QueryInterface(riid, ppv);
+}
+
+HRESULT CBandSiteMenu::_CreateBuiltInISFBand(UINT uID, REFIID riid, void** ppv)
+{
+    LPITEMIDLIST pidl;
+    HRESULT hr;
+
+    if (uID == IDM_TASKBAR_TOOLBARS_DESKTOP)
+        pidl = m_DesktopPidl;
+    else if (uID == IDM_TASKBAR_TOOLBARS_QUICKLAUNCH)
+        pidl = m_QLaunchPidl;
+
+    CComPtr<IShellFolderBand> pISFB;
+    hr = CISFBand_CreateInstance(IID_IShellFolderBand, (PVOID*)&pISFB);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    hr = pISFB->InitializeSFB(NULL, pidl);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    /* HACK! We shouldn't pass ISFB_STATE_QLINKSMODE and CISFBand shouldn't handle it! */
+    if (uID == IDM_TASKBAR_TOOLBARS_QUICKLAUNCH)
+    {
+        BANDINFOSFB bisfb = {ISFB_MASK_STATE, ISFB_STATE_QLINKSMODE, ISFB_STATE_QLINKSMODE};
+        pISFB->SetBandInfoSFB(&bisfb);
+    }
+    
+    return pISFB->QueryInterface(riid, ppv);
+}
+
+HRESULT CBandSiteMenu::_AddISFBandToMenu(HMENU hmenu, UINT indexMenu, UINT idCmdFirst, UINT idCmdLast, IUnknown* pBand, DWORD dwBandID, UINT *newMenuId)
+{
+    CComPtr<IShellFolderBand> psfb;
+    HRESULT hr = pBand->QueryInterface(IID_PPV_ARG(IShellFolderBand, &psfb));
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    BANDINFOSFB bi = {ISFB_MASK_IDLIST};
+    hr = psfb->GetBandInfoSFB(&bi);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    if (!bi.pidl)
+        return E_OUTOFMEMORY;
+
+    WCHAR buffer[MAX_PATH];
+    hr = ILGetDisplayNameEx(NULL, bi.pidl, buffer, ILGDN_INFOLDER) ? S_OK : E_FAIL;
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    UINT id = idCmdFirst + DSA_GetItemCount(m_menuDsa) + 0x10 + dwBandID;
+    if (id >= idCmdLast)
+        return E_FAIL;
+
+    *newMenuId = id;
+    InsertMenu(hmenu, indexMenu, MF_BYPOSITION, id, buffer);
+    return S_OK;
+}
+
+UINT CBandSiteMenu::_GetMenuIdFromISFBand(IUnknown *pBand)
+{
+    UINT ret = UINT_MAX;
+
+    CComPtr<IShellFolderBand> psfb;
+    HRESULT hr = pBand->QueryInterface(IID_PPV_ARG(IShellFolderBand, &psfb));
+    if (FAILED_UNEXPECTEDLY(hr))
+        return ret;
+
+    BANDINFOSFB bi = {ISFB_MASK_IDLIST};
+    hr = psfb->GetBandInfoSFB(&bi);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return ret;
+
+    LPITEMIDLIST pidl = bi.pidl;
+    if (!pidl)
+        return ret;
+
+    if (pidl->mkid.cb == 0)
```

Comment:
_ILIsDesktop or such?

Reply:
We need to create a helper header and lib at some point for some shared code in various places in the shell including this one.